### PR TITLE
Only check for approximate result for gdk-pixbuf decoding

### PR DESCRIPTION
**Pull request by Schmiddiii**
The results given by the gdk-pixbuf implementation seems to be flaky, probably due to their implementation not being stable between versions or not deterministic. Instead, require only that the images "look similar". I have coded the "similarity" to be based on the average error on the pixel values and a peak error allowed, compared to the same image being encoded by blurhash-python (an official implementation of the blurhash algorithm). For that, I have also written a small script running blurhash-python on all `png` images in `data`.

### CODE REVIEW

**Comment by direc85** [comment]
That's unfortunate, if the pixels fluctuate around instead of being deterministic...

There's [image-compare](https://crates.io/crates/image-compare) that could also be used. That way the allowed difference would be easier to figure out, adjust, and print out if the difference is too great.

**Comment by Schmiddiii** [comment]
`image-compare` indeed looks interesting, but I don't think we need to add a new dependency for an average of `0` pixel errors and maximum of `1` bitvalue off the original (I chose to use bigger error margins when checking, but those were the values I observed). We could use that if we notice bigger issues with some pictures though.

File: src/lib.rs [hunk 1]
```diff
@@ -450,8 +447,38 @@ mod tests {
...
+        }
+
+        let average_error = aggregated_error / result.len();
+
+        if average_error > MAX_AVERAGE_ERROR || peak_error > MAX_PEAK_ERROR {
+            panic!(
+                "images do not look similar: average error {}; peak error {}",
+                average_error, peak_error
+            );
+        }
```
Comment (rubdos) [inline src/lib.rs:482]:
```suggestion
        assert!(average_error <= MAX_AVERAGE_ERROR, "images do not look similar. average error {average_error} > {MAX_AVERAGE_ERROR}");
        assert!(peak_error <= MAX_PEAK_ERROR, "images do not look similar. peak error {peak_error} > {MAX_PEAK_ERROR}");
```

I think asserts are a bit more clear in this context. Maybe the same above.

File: src/lib.rs [hunk 2]
```diff
@@ -450,8 +447,38 @@ mod tests {
...
+        const MAX_AVERAGE_ERROR: usize = 1;
+        const MAX_PEAK_ERROR: usize = 8;
+
+        if result.len() != target.len() {
+            panic!(
+                "images do not have the same shape: {} vs {}",
+                result.len(),
+                target.len()
+            );
+        }
```
Comment (rubdos) [inline src/lib.rs:465]:
```suggestion
        assert_eq!(result.len(), target.len(), "images do not have the same shape: {} vs {}", result.len(), target.len());
```

**Review by rubdos — COMMENTED** [review]
Looks pretty good to me. Maybe worth adding a small comment on top of the blur.py to document why it exists.

**Comment by rubdos** [comment]
If this pops up again, we can also implement an [SSIM](https://en.wikipedia.org/wiki/Structural_similarity_index_measure). That should tackle this for good.

Otoh, we're probably the fastest and the best tested blurhash lib around at this point...

**Comment by Schmiddiii** [comment]
Needed to change the format strings to be `{}` instead of `{...}` as yours created warnings with Rust.

> If this pops up again, we can also implement an [SSIM](https://en.wikipedia.org/wiki/Structural_similarity_index_measure). That should tackle this for good.

Never heared of SSIM before, but sounds interesting.

**Comment by rubdos** [comment]
> Needed to change the format strings to be {} instead of {...} as yours created warnings with Rust.

Oh, sure, we still have 1.52 on support I suppose...

> Never heared of SSIM before, but sounds interesting.

I kinda did a master's in multimedia processing, so I heard these things once or twice ;-)

**Review by rubdos — APPROVED** [review]

**Pull request merged**